clients/drcachesim/tracer/tracer.cpp

```diff
@@ -720,69 +790,15 @@ instrument_clean_call(void *drcontext, instrlist_t *ilist, instr_t *where,
-                                  OPND_CREATE_MEMPTR(reg_global, 0)));
-        MINSERT(ilist, where, INSTR_CREATE_dmb(drcontext, OPND_CREATE_INT(DR_DMB_ISH)));
-#    endif
-#else
-        MINSERT(ilist, where,
-                XINST_CREATE_load(drcontext, opnd_create_reg(reg_global),
-                                  OPND_CREATE_ABSMEM(&tracing_window, OPSZ_PTR)));
-#endif
-        dr_insert_read_raw_tls(drcontext, ilist, where, tls_seg,
-                               tls_offs + sizeof(void *) * MEMTRACE_TLS_OFFS_WINDOW,
-                               reg_mine);
-#ifdef AARCHXX
-        MINSERT(ilist, where,
-                XINST_CREATE_sub(drcontext, opnd_create_reg(reg_mine),
-                                 opnd_create_reg(reg_global)));
-#elif defined(RISCV64)
-        /* FIXME i#3544: Not implemented */
-        DR_ASSERT_MSG(false, "Not implemented on RISC-V");
-#else
-        // Our version of a flags-free reg-reg subtraction: 1's complement one reg
-        // plus 1 and then add using base+index of LEA.
-        MINSERT(ilist, where, INSTR_CREATE_not(drcontext, opnd_create_reg(reg_global)));
-        MINSERT(ilist, where,
-                INSTR_CREATE_lea(drcontext, opnd_create_reg(reg_global),
-                                 OPND_CREATE_MEM_lea(reg_global, DR_REG_NULL, 0, 1)));
-        MINSERT(ilist, where,
-                INSTR_CREATE_lea(drcontext, opnd_create_reg(reg_mine),
-                                 OPND_CREATE_MEM_lea(reg_mine, reg_global, 1, 0)));
-#endif
-        // To avoid writing a 0 on top of the redzone, we read the buffer value and add
-        // that to the local ("mine") window minus the global window.  The redzone is
-        // -1, so if we do mine minus global which will always be non-positive, we'll
-        // never write 0 for a redzone slot (and thus possibly overflowing).
-        MINSERT(ilist, where,
-                XINST_CREATE_load(drcontext, opnd_create_reg(reg_global),
-                                  OPND_CREATE_MEMPTR(reg_ptr, 0)));
-        MINSERT(ilist, where,
-                XINST_CREATE_add(drcontext, opnd_create_reg(reg_mine),
-                                 opnd_create_reg(reg_global)));
-        MINSERT(ilist, where,
-                XINST_CREATE_store(drcontext, OPND_CREATE_MEMPTR(reg_ptr, 0),
-                                   opnd_create_reg(reg_mine)));
-        if (drreg_unreserve_register(drcontext, ilist, where, reg_global) !=
-                DRREG_SUCCESS ||
-            drreg_unreserve_register(drcontext, ilist, where, reg_mine) != DRREG_SUCCESS)
-            FATAL("Fatal error: failed to unreserve scratch reg.\n");
+        insert_mode_comparison(drcontext, ilist, where, reg_ptr, &tracing_window,
+                               MEMTRACE_TLS_OFFS_WINDOW);
+    }
+
+    if (op_L0_filter_until_instrs.get_value()) {
+        // Force a clean call when the tracing mode changes, so that the other
+        // threads can update their traces appropriately.
```

Comment:
I found the wording confusing: please reword the comment as it sounds like it is this thread that changed the mode when really it's the opposite: this is one of the "other threads".

Reply:
Done.